File: Repo/opencog/atomspace/Go/atomspace.go
// Package atomspace provides a Go implementation of the OpenCog AtomSpace
// knowledge representation system.
//
// The AtomSpace is a hypergraph database for storing and manipulating
// knowledge in the form of Atoms (Nodes and Links).
package atomspace

import (
	"fmt"
	"sync"
)

// AtomType represents the type of an Atom
type AtomType string

const (
	// Node types
	ConceptNode   AtomType = "ConceptNode"
	PredicateNode AtomType = "PredicateNode"
	VariableNode  AtomType = "VariableNode"
	NumberNode    AtomType = "NumberNode"
	
	// Link types
	InheritanceLink AtomType = "InheritanceLink"
	SimilarityLink  AtomType = "SimilarityLink"
	EvaluationLink  AtomType = "EvaluationLink"
	ListLink        AtomType = "ListLink"
	AndLink         AtomType = "AndLink"
	OrLink          AtomType = "OrLink"
)

// TruthValue represents the truth value of an Atom
type TruthValue struct {
	Strength   float64 // [0.0, 1.0]
	Confidence float64 // [0.0, 1.0]
}

// NewTruthValue creates a new TruthValue with validation
func NewTruthValue(strength, confidence float64) TruthValue {
	if strength < 0.0 {
		strength = 0.0
	} else if strength > 1.0 {
		strength = 1.0
	}
	
	if confidence < 0.0 {
		confidence = 0.0
	} else if confidence > 1.0 {
		confidence = 1.0
	}
	
	return TruthValue{Strength: strength, Confidence: confidence}
}

// DefaultTruthValue returns a default truth value
func DefaultTruthValue() TruthValue {
	return TruthValue{Strength: 1.0, Confidence: 1.0}
}

// Atom is the base interface for all atoms in the AtomSpace
type Atom interface {
	GetType() AtomType
	GetHandle() Handle
	GetTruthValue() TruthValue
	SetTruthValue(tv TruthValue)
	String() string
}

// Handle is a unique identifier for an Atom
type Handle uint64

// Node represents a node in the AtomSpace
type Node struct {
	handle     Handle
	atomType   AtomType
	name       string
	truthValue TruthValue
	mu         sync.RWMutex
}

// NewNode creates a new Node
func NewNode(atomType AtomType, name string) *Node {
	return &Node{
		atomType:   atomType,
		name:       name,
		truthValue: DefaultTruthValue(),
	}
}

// GetType returns the atom type
func (n *Node) GetType() AtomType {
	return n.atomType
}

// GetHandle returns the atom handle
func (n *Node) GetHandle() Handle {
	return n.handle
}

// GetName returns the node name
func (n *Node) GetName() string {
	return n.name
}

// GetTruthValue returns the truth value
func (n *Node) GetTruthValue() TruthValue {
	n.mu.RLock()
	defer n.mu.RUnlock()
	return n.truthValue
}

// SetTruthValue sets the truth value
func (n *Node) SetTruthValue(tv TruthValue) {
	n.mu.Lock()
	defer n.mu.Unlock()
	n.truthValue = tv
}

// String returns a string representation
func (n *Node) String() string {
	return fmt.Sprintf("(%s \"%s\")", n.atomType, n.name)
}

// Link represents a link between atoms in the AtomSpace
type Link struct {
	handle     Handle
	atomType   AtomType
	outgoing   []Atom
	truthValue TruthValue
	mu         sync.RWMutex
}

// NewLink creates a new Link
func NewLink(atomType AtomType, outgoing []Atom) *Link {
	return &Link{
		atomType:   atomType,
		outgoing:   outgoing,
		truthValue: DefaultTruthValue(),
	}
}

// GetType returns the atom type
func (l *Link) GetType() AtomType {
	return l.atomType
}

// GetHandle returns the atom handle
func (l *Link) GetHandle() Handle {
	return l.handle
}

// GetOutgoing returns the outgoing set
func (l *Link) GetOutgoing() []Atom {
	return l.outgoing
}

// GetTruthValue returns the truth value
func (l *Link) GetTruthValue() TruthValue {
	l.mu.RLock()
	defer l.mu.RUnlock()
	return l.truthValue
}

// SetTruthValue sets the truth value
func (l *Link) SetTruthValue(tv TruthValue) {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.truthValue = tv
}

// String returns a string representation
func (l *Link) String() string {
	outStr := ""
	for i, atom := range l.outgoing {
		if i > 0 {
			outStr += " "
		}
		outStr += atom.String()
	}
	return fmt.Sprintf("(%s %s)", l.atomType, outStr)
}

// AtomSpace is the main knowledge representation database
type AtomSpace struct {
	atoms       map[Handle]Atom
	nodeIndex   map[string]map[string]Handle // type -> name -> handle
	nextHandle  Handle
	mu          sync.RWMutex
}

// NewAtomSpace creates a new AtomSpace
func NewAtomSpace() *AtomSpace {
	return &AtomSpace{
		atoms:      make(map[Handle]Atom),
		nodeIndex:  make(map[string]map[string]Handle),
		nextHandle: 1,
	}
}

// AddNode adds a node to the AtomSpace
func (as *AtomSpace) AddNode(atomType AtomType, name string) *Node {
	as.mu.Lock()
	defer as.mu.Unlock()
	
	// Check if node already exists
	if typeMap, exists := as.nodeIndex[string(atomType)]; exists {
		if handle, found := typeMap[name]; found {
			if node, ok := as.atoms[handle].(*Node); ok {
				return node
			}
		}
	}
	
	// Create new node
	node := NewNode(atomType, name)
	handle := as.nextHandle
	as.nextHandle++
	
	node.handle = handle
	as.atoms[handle] = node
	
	// Update index
	if as.nodeIndex[string(atomType)] == nil {
		as.nodeIndex[string(atomType)] = make(map[string]Handle)
	}
	as.nodeIndex[string(atomType)][name] = handle
	
	return node
}

// AddLink adds a link to the AtomSpace
func (as *AtomSpace) AddLink(atomType AtomType, outgoing []Atom) *Link {
	as.mu.Lock()
	defer as.mu.Unlock()
	
	// TODO: Check for duplicate links
	
	link := NewLink(atomType, outgoing)
	handle := as.nextHandle
	as.nextHandle++
	
	link.handle = handle
	as.atoms[handle] = link
	
	return link
}

// GetAtom retrieves an atom by handle
func (as *AtomSpace) GetAtom(handle Handle) Atom {
	as.mu.RLock()
	defer as.mu.RUnlock()
	return as.atoms[handle]
}

// GetNode retrieves a node by type and name
func (as *AtomSpace) GetNode(atomType AtomType, name string) *Node {
	as.mu.RLock()
	defer as.mu.RUnlock()
	
	if typeMap, exists := as.nodeIndex[string(atomType)]; exists {
		if handle, found := typeMap[name]; found {
			if node, ok := as.atoms[handle].(*Node); ok {
				return node
			}
		}
	}
	return nil
}

// RemoveAtom removes an atom from the AtomSpace
func (as *AtomSpace) RemoveAtom(handle Handle) bool {
	as.mu.Lock()
	defer as.mu.Unlock()
	
	atom, exists := as.atoms[handle]
	if !exists {
		return false
	}
	
	// Remove from index if it's a node
	if node, ok := atom.(*Node); ok {
		if typeMap, exists := as.nodeIndex[string(node.atomType)]; exists {
			delete(typeMap, node.name)
		}
	}
	
	delete(as.atoms, handle)
	return true
}

// GetAllAtoms returns all atoms in the AtomSpace
func (as *AtomSpace) GetAllAtoms() []Atom {
	as.mu.RLock()
	defer as.mu.RUnlock()
	
	atoms := make([]Atom, 0, len(as.atoms))
	for _, atom := range as.atoms {
		atoms = append(atoms, atom)
	}
	return atoms
}

// GetAtomsByType returns all atoms of a specific type
func (as *AtomSpace) GetAtomsByType(atomType AtomType) []Atom {
	as.mu.RLock()
	defer as.mu.RUnlock()
	
	atoms := make([]Atom, 0)
	for _, atom := range as.atoms {
		if atom.GetType() == atomType {
			atoms = append(atoms, atom)
		}
	}
	return atoms
}

// Size returns the number of atoms in the AtomSpace
func (as *AtomSpace) Size() int {
	as.mu.RLock()
	defer as.mu.RUnlock()
	return len(as.atoms)
}

// Clear removes all atoms from the AtomSpace
func (as *AtomSpace) Clear() {
	as.mu.Lock()
	defer as.mu.Unlock()
	
	as.atoms = make(map[Handle]Atom)
	as.nodeIndex = make(map[string]map[string]Handle)
	as.nextHandle = 1
}

File: Repo/opencog/atomspace/Julia/atomspace.jl
#!/usr/bin/env julia
#=
atomspace.jl

OpenCog AtomSpace - Julia Implementation
Hypergraph-based knowledge representation system

This implementation demonstrates Julia's strengths:
- Multiple dispatch for polymorphic atom operations
- Type system with abstract types and concrete subtypes
- Parametric types for flexible data structures
- Performance through type stability
- Mathematical notation for truth values
- Broadcasting and vectorization
=#

using Printf

# ============================================================================
# ATOM TYPES AND CONSTANTS
# ============================================================================

# Atom types as enum-like constants
@enum AtomType begin
    # Node types
    CONCEPT_NODE = 1
    PREDICATE_NODE = 2
    VARIABLE_NODE = 3
    NUMBER_NODE = 4
    
    # Link types
    INHERITANCE_LINK = 5
    SIMILARITY_LINK = 6
    EVALUATION_LINK = 7
    LIST_LINK = 8
    AND_LINK = 9
    OR_LINK = 10
end

# Handle type for unique atom identification
const Handle = UInt64

# ============================================================================
# TRUTH VALUE SYSTEM
# ============================================================================

"""
    TruthValue

Represents probabilistic truth with strength and confidence values.
Demonstrates: Immutable struct, validation, mathematical operations
"""
struct TruthValue
    strength::Float64    # [0.0, 1.0]
    confidence::Float64  # [0.0, 1.0]
    
    # Inner constructor with validation
    function TruthValue(strength::Float64, confidence::Float64)
        s = clamp(strength, 0.0, 1.0)
        c = clamp(confidence, 0.0, 1.0)
        new(s, c)
    end
end

# Convenience constructors
TruthValue() = TruthValue(1.0, 1.0)
TruthValue(s::Real) = TruthValue(Float64(s), 1.0)

# Display method
Base.show(io::IO, tv::TruthValue) = print(io, "<$(tv.strength), $(tv.confidence)>")

# ============================================================================
# ABSTRACT ATOM HIERARCHY
# ============================================================================

"""
    Atom

Abstract base type for all atoms in the AtomSpace.
Demonstrates: Abstract type hierarchy, interface design
"""
abstract type Atom end

# Required interface methods (to be implemented by subtypes)
atom_type(a::Atom) = error("atom_type not implemented for $(typeof(a))")
handle(a::Atom) = error("handle not implemented for $(typeof(a))")
truth_value(a::Atom) = error("truth_value not implemented for $(typeof(a))")

# ============================================================================
# NODE IMPLEMENTATION
# ============================================================================

"""
    Node <: Atom

Represents a node in the hypergraph with a name.
Demonstrates: Concrete type, mutable fields, parametric types
"""
mutable struct Node <: Atom
    handle::Handle
    atom_type::AtomType
    name::String
    truth_value::TruthValue
end

# Constructor
Node(atom_type::AtomType, name::String) = Node(0, atom_type, name, TruthValue())
Node(atom_type::AtomType, name::String, tv::TruthValue) = Node(0, atom_type, name, tv)

# Interface implementations using multiple dispatch
atom_type(n::Node) = n.atom_type
handle(n::Node) = n.handle
truth_value(n::Node) = n.truth_value

# Setters
set_handle!(n::Node, h::Handle) = (n.handle = h)
set_truth_value!(n::Node, tv::TruthValue) = (n.truth_value = tv)

# Display
Base.show(io::IO, n::Node) = print(io, "($(n.atom_type) \"$(n.name)\")")

# Equality based on type and name
Base.:(==)(n1::Node, n2::Node) = n1.atom_type == n2.atom_type && n1.name == n2.name
Base.hash(n::Node, h::UInt) = hash((n.atom_type, n.name), h)

# ============================================================================
# LINK IMPLEMENTATION
# ============================================================================

"""
    Link <: Atom

Represents a link between atoms in the hypergraph.
Demonstrates: Vector fields, recursive structures
"""
mutable struct Link <: Atom
    handle::Handle
    atom_type::AtomType
    outgoing::Vector{Atom}
    truth_value::TruthValue
end

# Constructor
Link(atom_type::AtomType, outgoing::Vector{Atom}) = Link(0, atom_type, outgoing, TruthValue())
Link(atom_type::AtomType, outgoing::Vector{Atom}, tv::TruthValue) = Link(0, atom_type, outgoing, tv)

# Interface implementations
atom_type(l::Link) = l.atom_type
handle(l::Link) = l.handle
truth_value(l::Link) = l.truth_value

# Setters
set_handle!(l::Link, h::Handle) = (l.handle = h)
set_truth_value!(l::Link, tv::TruthValue) = (l.truth_value = tv)

# Accessors
outgoing(l::Link) = l.outgoing
arity(l::Link) = length(l.outgoing)

# Display
function Base.show(io::IO, l::Link)
    out_str = join([string(a) for a in l.outgoing], " ")
    print(io, "($(l.atom_type) $out_str)")
end

# ============================================================================
# ATOMSPACE IMPLEMENTATION
# ============================================================================

"""
    AtomSpace

Main hypergraph knowledge representation database.
Demonstrates: Mutable struct with complex state, thread-safe operations
"""
mutable struct AtomSpace
    atoms::Dict{Handle, Atom}
    node_index::Dict{Tuple{AtomType, String}, Handle}
    next_handle::Handle
    lock::ReentrantLock
    
    function AtomSpace()
        new(
            Dict{Handle, Atom}(),
            Dict{Tuple{AtomType, String}, Handle}(),
            1,
            ReentrantLock()
        )
    end
end

# ============================================================================
# ATOMSPACE OPERATIONS
# ============================================================================

"""
    add_node!(as::AtomSpace, atom_type::AtomType, name::String) -> Node

Add a node to the AtomSpace. Returns existing node if already present.
Demonstrates: Multiple dispatch, thread safety, idempotent operations
"""
function add_node!(as::AtomSpace, atom_type::AtomType, name::String)
    lock(as.lock) do
        # Check if node already exists
        key = (atom_type, name)
        if haskey(as.node_index, key)
            handle = as.node_index[key]
            return as.atoms[handle]::Node
        end
        
        # Create new node
        node = Node(atom_type, name)
        h = as.next_handle
        as.next_handle += 1
        
        set_handle!(node, h)
        as.atoms[h] = node
        as.node_index[key] = h
        
        return node
    end
end

# Overload with truth value
function add_node!(as::AtomSpace, atom_type::AtomType, name::String, tv::TruthValue)
    node = add_node!(as, atom_type, name)
    set_truth_value!(node, tv)
    return node
end

"""
    add_link!(as::AtomSpace, atom_type::AtomType, outgoing::Vector{Atom}) -> Link

Add a link to the AtomSpace.
Demonstrates: Generic programming, type constraints
"""
function add_link!(as::AtomSpace, atom_type::AtomType, outgoing::Vector{Atom})
    lock(as.lock) do
        # TODO: Check for duplicate links
        
        link = Link(atom_type, outgoing)
        h = as.next_handle
        as.next_handle += 1
        
        set_handle!(link, h)
        as.atoms[h] = link
        
        return link
    end
end

# Overload with truth value
function add_link!(as::AtomSpace, atom_type::AtomType, outgoing::Vector{Atom}, tv::TruthValue)
    link = add_link!(as, atom_type, outgoing)
    set_truth_value!(link, tv)
    return link
end

"""
    get_atom(as::AtomSpace, handle::Handle) -> Union{Atom, Nothing}

Retrieve an atom by its handle.
"""
function get_atom(as::AtomSpace, handle::Handle)
    lock(as.lock) do
        get(as.atoms, handle, nothing)
    end
end

"""
    get_node(as::AtomSpace, atom_type::AtomType, name::String) -> Union{Node, Nothing}

Retrieve a node by type and name.
"""
function get_node(as::AtomSpace, atom_type::AtomType, name::String)
    lock(as.lock) do
        key = (atom_type, name)
        if haskey(as.node_index, key)
            handle = as.node_index[key]
            return as.atoms[handle]::Node
        end
        return nothing
    end
end

"""
    remove_atom!(as::AtomSpace, handle::Handle) -> Bool

Remove an atom from the AtomSpace.
"""
function remove_atom!(as::AtomSpace, handle::Handle)
    lock(as.lock) do
        if !haskey(as.atoms, handle)
            return false
        end
        
        atom = as.atoms[handle]
        
        # Remove from node index if it's a node
        if atom isa Node
            key = (atom.atom_type, atom.name)
            delete!(as.node_index, key)
        end
        
        delete!(as.atoms, handle)
        return true
    end
end

"""
    get_all_atoms(as::AtomSpace) -> Vector{Atom}

Return all atoms in the AtomSpace.
"""
function get_all_atoms(as::AtomSpace)
    lock(as.lock) do
        collect(values(as.atoms))
    end
end

"""
    get_atoms_by_type(as::AtomSpace, atom_type::AtomType) -> Vector{Atom}

Return all atoms of a specific type.
Demonstrates: Filtering, type checking
"""
function get_atoms_by_type(as::AtomSpace, atom_type::AtomType)
    lock(as.lock) do
        filter(a -> atom_type(a) == atom_type, collect(values(as.atoms)))
    end
end

"""
    size(as::AtomSpace) -> Int

Return the number of atoms in the AtomSpace.
"""
Base.size(as::AtomSpace) = lock(() -> length(as.atoms), as.lock)
Base.length(as::AtomSpace) = size(as)

"""
    isempty(as::AtomSpace) -> Bool

Check if the AtomSpace is empty.
"""
Base.isempty(as::AtomSpace) = lock(() -> isempty(as.atoms), as.lock)

"""
    empty!(as::AtomSpace)

Remove all atoms from the AtomSpace.
"""
function Base.empty!(as::AtomSpace)
    lock(as.lock) do
        empty!(as.atoms)
        empty!(as.node_index)
        as.next_handle = 1
    end
end

# Membership test
Base.in(atom::Atom, as::AtomSpace) = lock(() -> handle(atom) in keys(as.atoms), as.lock)

# ============================================================================
# CONVENIENCE FUNCTIONS
# ============================================================================

"""
    concept_node(as::AtomSpace, name::String) -> Node

Create a concept node. Convenience function.
"""
concept_node(as::AtomSpace, name::String) = add_node!(as, CONCEPT_NODE, name)

"""
    inheritance_link(as::AtomSpace, from::Atom, to::Atom) -> Link

Create an inheritance link. Convenience function.
"""
inheritance_link(as::AtomSpace, from::Atom, to::Atom) = add_link!(as, INHERITANCE_LINK, [from, to])

# ============================================================================
# DEMONSTRATION FUNCTION
# ============================================================================

"""
    demo()

Demonstrate AtomSpace functionality.
"""
function demo()
    println("="^70)
    println("OpenCog AtomSpace - Julia Implementation Demo")
    println("Showcasing hypergraph knowledge representation")
    println("="^70)
    println()
    
    # Create AtomSpace
    println("1. Creating AtomSpace")
    println("-"^70)
    as = AtomSpace()
    println("  Created empty AtomSpace: size = $(size(as))")
    println()
    
    # Add nodes
    println("2. Adding Nodes")
    println("-"^70)
    human = add_node!(as, CONCEPT_NODE, "human")
    mortal = add_node!(as, CONCEPT_NODE, "mortal")
    socrates = add_node!(as, CONCEPT_NODE, "Socrates")
    
    println("  Added nodes:")
    println("    $human")
    println("    $mortal")
    println("    $socrates")
    println("  AtomSpace size: $(size(as))")
    println()
    
    # Add links
    println("3. Adding Links")
    println("-"^70)
    link1 = add_link!(as, INHERITANCE_LINK, [human, mortal])
    link2 = add_link!(as, INHERITANCE_LINK, [socrates, human])
    
    println("  Added links:")
    println("    $link1")
    println("    $link2")
    println("  AtomSpace size: $(size(as))")
    println()
    
    # Truth values
    println("4. Truth Values")
    println("-"^70)
    tv_strong = TruthValue(0.95, 0.90)
    tv_weak = TruthValue(0.60, 0.50)
    
    set_truth_value!(link1, tv_strong)
    set_truth_value!(link2, tv_weak)
    
    println("  Link 1 TV: $(truth_value(link1))")
    println("  Link 2 TV: $(truth_value(link2))")
    println()
    
    # Query operations
    println("5. Query Operations")
    println("-"^70)
    retrieved = get_node(as, CONCEPT_NODE, "human")
    println("  Retrieved node: $retrieved")
    
    concepts = get_atoms_by_type(as, CONCEPT_NODE)
    println("  Concept nodes ($(length(concepts))):")
    for c in concepts
        println("    $c")
    end
    
    links = get_atoms_by_type(as, INHERITANCE_LINK)
    println("  Inheritance links ($(length(links))):")
    for l in links
        println("    $l")
    end
    println()
    
    # Performance demonstration
    println("6. Performance - Batch Operations")
    println("-"^70)
    @time begin
        for i in 1:1000
            add_node!(as, CONCEPT_NODE, "concept_$i")
        end
    end
    println("  Created 1000 nodes")
    println("  Final size: $(size(as))")
    println()
    
    println("="^70)
    println("Julia AtomSpace strengths demonstrated:")
    println("  ✓ Multiple dispatch for polymorphic operations")
    println("  ✓ Type system with abstract and concrete types")
    println("  ✓ Thread-safe operations with locks")
    println("  ✓ Mathematical notation for truth values")
    println("  ✓ Performance through type stability")
    println("  ✓ Rich standard library integration")
    println("="^70)
end

# Run demo if executed as script
if abspath(PROGRAM_FILE) == @__FILE__
    demo()
end
